### ISSUE TITLE
MAINT: optimize.shgo: returns incorrect solution to Rosenbrock problem

### ISSUE THREAD

**Issue by Kvothe-The-Bloodless** [scipy/scipy#14533]
My issue is about optimize.shgo. It gives an unexpected TypeError.

When running the following

    from scipy.optimize import rosen, rosen_der, rosen_hess
    bounds = [(0,1.6), (0, 1.6), (0, 1.4), (0, 1.4), (0, 1.4)]
    result = scipy.optimize.shgo(rosen, bounds, options={'jac':rosen_der,'hess':rosen_hess})

I get

    TypeError: _minimize_slsqp() got multiple values for argument 'jac'

I believe that `jac` is correctly specified here (see [this documentation][1]). Did I make a mistake or is there a bug here?

  [1]: https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.shgo.html#scipy.optimize.shgo

#### Scipy/Numpy/Python version information:
```
1.6.0 1.20.0 sys.version_info(major=3, minor=8, micro=5, releaselevel='final', serial=0)
```

**Comment by Stefan-Endres** [comment]
Thank you for reporting @Kvothe-The-Bloodless. This is a duplicate of #13547 and will be fixed in #14013.

**Comment by Kvothe-The-Bloodless** [comment]
> Thank you for reporting @Kvothe-The-Bloodless. This is a duplicate of #13547 and will be fixed in #14013.

Thank you. Are you sure? The issue described there seems to have been that the gradient could not be provided with 'jac': True. Now it seems that no method of providing the jacobian is available at all. Also https://github.com/scipy/scipy/issues/12963 suggest that the problem is in the documentation and that `jac`,`hess` should not be provided through options but through `minimizer_kwargs`.

Is there currently any functioning way to provide the jacobian and the hessian?

**Comment by Stefan-Endres** [comment]
No I am wrong, it is a separate issue, which I thought would also fixed with this diff https://github.com/scipy/scipy/pull/14013/files#diff-d87298173e00845c0e28dfa30ed635c9d83f174bc2ad35a88ba6af9efcc62922L459-R500, having just tested it it, it does not appear to be working with the fix to update the `minimizer_kwargs` dict with the solver options from `options` correctly. When the `jac` and `hess` arguments are present in the global `options` as in the documentation it should also pass back to `minimizer_kwargs`, but without duplication.

>Is there currently any functioning way to provide the jacobian and the hessian?

The [unittest `test_7_3_minkwargs`](https://github.com/scipy/scipy/blob/master/scipy/optimize/tests/test__shgo.py#L577-L594) is passing so `jac` and `hess` is working for all the other sovlers when being passed through `shgo`, but the `COBYLA`, `SLSQP` sovlers are excluded here because they do not support `jac` (`COBYLA`) or `hess` (both `COBYLA` and `SLSQP`) arguments. In the future after finding the cause of the issue a new `jac` test for `SLSQP` is needed, since ` _minimize_slsqp() ` can ignore `hess` it would be simple matter of adding `SLSQP` after the fix.


Currently this only works without error for `SLSQP` by directly passing the `minimizer_kwargs` to the solver. Example:

```Python
from scipy.optimize import rosen, rosen_der, rosen_hess
bounds = [(0, 1.6), (0, 1.6), (0, 1.4), (0, 1.4), (0, 1.4)]
minimizer_kwargs['options'] = {'jac': rosen_der, 'hess': rosen_hess}
result = scipy.optimize.shgo(rosen, bounds,
                             minimizer_kwargs={'jac': rosen_der, 'hess': rosen_hess})
```
but passing it through the `options` argument is also supposed to work as per the documentation, please keep the issue open.

**Comment by Kvothe-The-Bloodless** [comment]
Thanks a lot! That seems to do the trick! (After adding `minimizer_kwargs={}` at the top.)

So the double use of `minimizer_kwargs` here is (currently) necessary and should just contain the same options?

(And if you don't mind me asking a slightly off-topic question here: What method would you recommend to specify if I have access to a free gradient and hessian? And if I understand correctly from another topic the global optimization part also uses these? 
I have seen that shgo was mainly meant for derivative-free optimization but I am still trying whether it is also useful for my non-convex optimization problem where I have access to a free gradient and hessian.)

**Comment by Stefan-Endres** [comment]
Yes, apologies for the confusing documentation, `minimizer_kwargs` which is also passed to `scipy.optimize.minimize` is necessary since the global `options` should also pass to `scipy.optimize.minimize` (this was done to streamline the documentation of the global solvers with `scipy.optimize.minimize`), but produces the bug as demonstrated in this issue.

>What method would you recommend to specify if I have access to a free gradient and hessian? 

I am not sure what is meant by "free gradient" here, do you mean you have it cached or it can be computed cheaply? 

>And if I understand correctly from another topic the global optimization part also uses these?
I have seen that shgo was mainly meant for derivative-free optimization but I am still trying whether it is also useful for my non-convex optimization problem where I have access to a free gradient and hessian.

Yes, `shgo` is mainly developed for gradient free (no known Jacobian or difficult to compute) and for problems you need more than one global solution (or you also need to explore other local minima in the search space of your problem). I am not an expert on the local minimization routines, but personally I have had by far the most succes with `method=’trust-krylov’` when I can suppy both a `jac` and `hess` to the problem both for accuracy and performance. 

The global routine does not use the Jacobian or Hessian, but for example when it is possible to compute Lipschitz constants and then exclude local domains (similar to DIRECT/DISIMPL-V) from further exploration (but this has to be done using the `SHGO.iterate()` and is only experimental so it is not documented or published anywhere yet; I would be happy to elaborate/send code if you are interested). The global routine does greatly increase in computational performance if the `symmetry` of the problem is known because the sampling domain size can be reduced up to maximum of `O(dim!)` times.

**Comment by Kvothe-The-Bloodless** [comment]
Thanks for the advice! 

Yes, I should have been more clear in my language. For my problem both the gradient and the hessian can be computed nearly free by an analytic trick once the function has been evaluated at a point. On the other hand each function evaluation is very costly. Basically all other cost of the optimization algorithm will pale in comparison. So I am looking to use some Newton-like solver. The problem I am encountering though is non-convexity. This hurts in two ways: 1.) Newton's type optimization algorithms don't work well because there is no obvious newton-step because the Hessian is not positive definite (of course there are some workarounds to force the Hessian to be positive definite such as those used in 'trust-constr', 'Newton-CG','BFGS' ).  2.) Getting attracted by a "wrong" minimum.

I am interested in finding minima in a certain "interesting" region say 0<x_1<0.5, 0<x_2<1.5,... I expect there to only be one minimum in this region, but it would nice to be more confident about this, for which I think this algorithm should help.  There is a sizable basin of attraction around the expected position of the "interesting" minimum, but there is a another "uninteresting" minimum (potentially smaller, i.e. the global minimum, but it seems there are settings to explore all candidates for local minima using shgo?). The exact barrier between the minima's is not know. The fact that the exact shape of the basin of attraction isn't know makes it difficult to  get attracted to the minimum of interest without stepping outside it. I managed in low 2 and 3 dimensional cases by choosing appropriate bounds and/or setting an appropriate initial step size by hand but this is a less than ideal solution. (The hessian does not provide a good scale/initial step size because it starts in an non-convex region.)

(I am looking to solve problems with 3-10 parameters.)

Another property of the function is that I know it is bounded from above by 1, and I roughly know it's minimal value to be say around -0.001. (This number is not exactly known, also the "uninteresting region" could attain a lower minimal value, perhaps even be unbounded from below.). This actually made me interested in what the option `f_min` exactly does? Would it be useful to me?

The symmetry option is potentially interesting, but for me there would be few parameters that would have a permutation symmetry and when they do I will probably be interested in different non-overlapping parameter regions for each of them, which should already prevent over counting. (I mean for example taking 0<x_1 < 1 < x_2 <2 where if x_1 and x_2 have a permutation symmetry taking this into account will not help since they are already bounded to different domains).

Anyway, thanks for the help. And sorry for boring you if you read all this ;). You seem really busy but if you do happen to have time to chat more I would be very interested. I'm a theoretical physicist that just happen to stumble on a problem that requires a tough optimization, so if I could talk to an expert like you I would be very grateful.

**Comment by Stefan-Endres** [comment]
> There is a sizable basin of attraction around the expected position of the "interesting" minimum, but there is a another "uninteresting" minimum (potentially smaller, i.e. the global minimum, but it seems there are settings to explore all candidates for local minima using shgo?).

Yes, by default shgo will explore all these minima. Once it has found the local minimum in a basin of attraction it will not search locally in that basin again (there is a theoretical advantage that it will not waste additional local searches in every subdomain that is a convex basin of attraction). If you keep iterating then it will explore all the local minima. Refining the complex globally has the guarantee that `shgo` will explore all the local minima until all are found (_iff_ the number of local minima are finite and the minima are strictly convex).

>The fact that the exact shape of the basin of attraction isn't know makes it difficult to get attracted to the minimum of interest without stepping outside it. I managed in low 2 and 3 dimensional cases by choosing appropriate bounds and/or setting an appropriate initial step size by hand but this is a less than ideal solution. (The hessian does not provide a good scale/initial step size because it starts in an non-convex region.)

This is also one of the reasons shgo was developed to supply bounds/constraints for every local basin of attraction it detects, it then constructs and passes the local constraints of that basin (based on the geometry known from the global sampling data in the current iteration) and keeps the local solver in the local basin of attraction (ex. in solutions thermodynamics this is important to find all phases in equilibrium). However, this of course only works in the limit where enough global exploration has been performed (by iterating more sampling points).

>(I am looking to solve problems with 3-10 parameters.)

This is perfect for `shgo`.

>This actually made me interested in what the option `f_min` exactly does? Would it be useful to me?

Yes, when you supply `f_min` the global algorithm will stop when a solution was found within a given `f_tol` of `f_min`, but if you would like to explore all minima with an objective function value around -0.001 or possibly lower value it would be better to not supply this `f_min` option.

>Anyway, thanks for the help. And sorry for boring you if you read all this ;). You seem really busy but if you do happen to have time to chat more I would be very interested. I'm a theoretical physicist that just happen to stumble on a problem that requires a tough optimization, so if I could talk to an expert like you I would be very grateful.

I would be very happy to talk more about your problem and help you further if needed. Unfortunately, I do not think GitHub is the appropriate forum to discuss use cases, but please feel free to contact me by e-mail (either [work](mailto:s.endres@iwt.uni-bremen.de)/[personal](mailto:stefan.c.endres@gmail.com) e-mail are both fine). Alternatively, we can post on the [scipy-user mailing lists](https://mail.python.org/mailman/listinfo/scipy-user) (which I believe is also appropriate), especially if you think the discussion could have value staying in the public domain.

**Comment by mdhaber** [comment]
@Stefan-Endres I believe this _was_ a duplicate of gh-12829 (different from above). This appears to have been fixed by gh-17140.

**Comment by mdhaber** [comment]
This was a duplicate as reported originally, but now the problem is that `shgo` returns the wrong answer yet reports success. If this is a pathological problem, that is OK, but if this is a reasonable enhancement request, let's open it. In any case, I don't think it should just implicitly be considered closed because the TypeError was fixed.

**Comment by Stefan-Endres** [comment]
I am not sure I follow, in the unittest `test_21_3_hess_options_rosen` in https://github.com/scipy/scipy/pull/16313 I get:

```
ref.fun = 1.073883798060949e-06
res.fun = 1.073883798060949e-06
```
tested with:
```python
assert_allclose(res.fun, ref.fun)
assert_allclose(res.x, ref.x, atol=1e-15)
```

The test is largely to ensure Hessian and Jacobian information can be used without errors and that it is possible to find the solution. Using the original code posted the main difference is in using the default `SLSQP` (which does not use Hessian information as clarified in https://github.com/scipy/scipy/issues/14533#issuecomment-892047109) rather than `Newton-CG` which is what you may have ran. If that is the case don't think anything discussed here can lead to an enchancement because it comes down to the performance of the local solvers  `SLSQP` vs.  `Newton-CG` which are being fed the same starting points/bounds from the global step of `shgo`.

**Comment by mdhaber** [comment]
OK, so the fact that the original code:
```python3
from scipy.optimize import rosen, rosen_der, rosen_hess
bounds = [(0,1.6), (0, 1.6), (0, 1.4), (0, 1.4), (0, 1.4)]
result = scipy.optimize.shgo(rosen, bounds, options={'jac':rosen_der,'hess':rosen_hess})
```
fails to find the correct solution is not considered to be a problem? You consider it to be SLSQP's fault, not that of SHGO?

**Comment by mdhaber** [comment]
@Stefan-Endres My concern was that the test, `test_21_3_hess_options_rosen`, has to pass special options in order for the correct global minimum to be found. The [original code](https://github.com/scipy/scipy/issues/14533#issuecomment-1407509186) in this issue returns the solution:

```
 message: Optimization terminated successfully.
 success: True
     fun: 1.1482239651668342
    funl: [ 1.148e+00]
       x: [ 8.000e-01  6.600e-01  4.483e-01  2.089e-01  4.362e-02]
      xl: [[ 8.000e-01  6.600e-01  4.483e-01  2.089e-01  4.362e-02]]
     nit: 1
    nfev: 67
   nlfev: 34
```

which is not very close to the global minimum.

To get the correct solution in `test_21_3_hess_options_rosen`, you had to specify `minimizer_kwargs = {'method': 'Newton-CG'}`. I see that changing the number `n` or changing `sampling_method` to `'sobol'` also works. So maybe this is just a funny case. I understand that we can't expect all problems to be solved correctly with default settings.

**Comment by mdhaber** [comment]
I have two more observations related to this issue, though: 

* Even in main, the original code does not raise a `TypeError`. We must have already fixed this problem.
* gh-16313 seems to introduce a problem when `iters > 1`

```python3
import scipy.optimize
from scipy.optimize import rosen, rosen_der, rosen_hess
bounds = [(0,1.6), (0, 1.6), (0, 1.4), (0, 1.4), (0, 1.4)]
%timeit scipy.optimize.shgo(rosen, bounds, options={'jac': rosen_der, 'hess': rosen_hess}, iters=2)
# main: 6.73 ms ± 362 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
# gh-16313: still waiting....
```

**Comment by Stefan-Endres** [comment]
>Even in main, the original code does not raise a TypeError. We must have already fixed this problem.

Yes, this was fixed in #17140 which we split off from the larger PR to get into 1.10.0.

>https://github.com/scipy/scipy/pull/16313 seems to introduce a problem when iters > 1

I found the same thing running it on the PR code, the bottleneck seems to be the new method to insert the minima into the complex between iterations (which resolved an upstream issue, not reported on gh-scipy, it is the intended behaviour of the method, but the first implementation was not working).

![image](https://user-images.githubusercontent.com/8235072/218251997-7af6e9b3-ed11-4802-a20e-1808d5bcdb37.png)

While I think it should be possible to replace this with much faster code, I will try to safely disable this and restore the old behaviour. Later I will open a smaller PR with benchmarks on higher dimensional problems.

**Comment by Stefan-Endres** [comment]
With [2791bdf](https://github.com/scipy/scipy/pull/16313/commits/2791bdfbe090b001d12296f4ca890ab014edb429) I get 223ms for the total call. It is slightly faster than the `main` branch due to at least not rerunning minimizers saved in the cache, but hopefully replaced the insertion with faster code in the future will drastically improve performance on problems with more expensive objective function calls.

@mdhaber if it's ok with you I would rather let this issue be closed with #16313 and ideally we can open a separate issue/PR where it's clear what is actually being requested with respect to the default settings, without being obscured by the technical details here. I think the central issue here is resolved by allowing the `jac` and `hess` to be passed properly to solve the problem without errors.

**Comment by mdhaber** [comment]
OK, for me, this is slower than in main
```
gh-16313: 15.4 ms ± 747 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
main: 9.43 ms ± 2.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
but I suppose this is something you intend to replace in the future, so we'll let it be for now

And as you wrote, this was really closed by gh-17140 rather than gh-16313, so I'll close it now.
